march_hardware_builder/include/march_hardware_builder/HardwareBuilder.h

```diff
@@ -0,0 +1,71 @@
+// Copyright 2019 Project March.
+
+#ifndef MARCH_IV_HARDWAREBUILDER_H
+#define MARCH_IV_HARDWAREBUILDER_H
+
+#include <yaml-cpp/yaml.h>
+
+#include <march_hardware/MarchRobot.h>
+#include <march_hardware/Joint.h>
+#include <march_hardware/Encoder.h>
+#include <march_hardware/IMotionCube.h>
+#include <march_hardware/TemperatureGES.h>
+
+#include <march_hardware_builder/AllowedRobot.h>
+
+/**
+ * @brief Creates a MarchRobot from a robot name, yamlPath, or a loaded .yaml config.
+ */
+class HardwareBuilder
+{
+private:
+  const std::vector<std::string> ENCODER_REQUIRED_KEYS =
+          { "resolution", "minPositionIU", "maxPositionIU", "zeroPositionIU", "safetyMarginRad" };
+  const std::vector<std::string> IMOTIONCUBE_REQUIRED_KEYS = { "slaveIndex", "encoder" };
+  const std::vector<std::string> TEMPERATUREGES_REQUIRED_KEYS = { "slaveIndex", "byteOffset" };
+  const std::vector<std::string> JOINT_REQUIRED_KEYS = {};
+
+public:
+  std::string yamlPath;
+  YAML::Node robotConfig;
+
+  /**
+   * @brief Initialise a HardwareBuilder with a robotName enumerator.
+   * @details Grabs the .yaml file associated with the robotName and loads it in the robotConfig.
+   */
+  explicit HardwareBuilder(AllowedRobot robotName);
+
+  /**
+   * @brief Initialise a HardwareBuilder with a path to a .yaml file.
+   */
+  explicit HardwareBuilder(std::string yamlPath);
+  HardwareBuilder();
+
+  /**
+   * @brief Create a MarchRobot from a yaml config.
+   */
+  march4cpp::MarchRobot createMarchRobot(YAML::Node marchRobotConfig);
+
+  /**
+   * @brief Create a MarchRobot, can only be used when a robotConfig is already loaded via the constructor.
+   */
+  march4cpp::MarchRobot createMarchRobot();
+
+  march4cpp::Joint createJoint(YAML::Node jointConfig, std::string jointName);
```

Comment:
Are the createX methods (apart from createMarchRobot) ever called from outside the HardwareBuilder? Otherwise maybe these should be private methods?

Reply:
Mainly for testing, I can look into the G_TEST friend class option

Reply:
The `FRIEND_TEST` appears to be a bad practice, the general consensus is that adding code for testing purposes is not idea. While these methods can be private at the moment, I believe they might be used publicly in the future.

Reply:
Allright